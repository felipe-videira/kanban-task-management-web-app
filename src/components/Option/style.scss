.option {
    border-radius: 50%;
    padding: 12% 12% calc(12% + (var(--size) * 0.04));
    width: 100%;
    height: 100%;
    border: none;
    box-shadow: inset 0px calc((var(--size) * 0.04) * -1) 0px 0px rgb(0 0 0 / 30%);
    cursor: pointer;

    &:active {
        box-shadow: none;
        padding: 12%;
        transform: translateY(calc(var(--size) * 0.04));
    }
    
    &__icon-container {
        box-shadow: inset 0px calc(var(--size) * 0.025) 0px 0px rgb(0 0 0 / 20%);
        background: #fff;
        border-radius: 50%;
        width: 100%;
        height: 100%;
        display: flex;
        align-items: center;
        justify-content: center;

        &:hover {
            background: #f0f0f0;
        }
    }

    &__icon {
        object-fit: contain;
        width: calc(var(--size) * 0.35);
        height: calc(var(--size) * 0.35);
    }
}

   

