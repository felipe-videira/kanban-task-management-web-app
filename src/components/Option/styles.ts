import styled from "styled-components/macro";

type StyleProps = {
  readonly size: number;
};

type ButtonProps = {
  readonly background: string | string[];
} & StyleProps;

export const OptionButton = styled.button.attrs<ButtonProps>((props) => {
  let background;

  if (Array.isArray(props.background) && props.background.length > 1) {
    background = `linear-gradient(${props.background.join(",")})`;
  } else if (Array.isArray(props.background)) {
    background = props.background[0];
  } else {
    background = props.background;
  }

  return {
    size: `${props.size * 0.04}px`,
    background,
  };
})<ButtonProps>`
  border-radius: 50%;
  padding: ${(props) => `12% 12% calc(12% + ${props.size})`};
  width: 100%;
  height: 100%;
  border: none;
  box-shadow: ${(props) => `inset 0px -${props.size} 0px 0px rgb(0 0 0 / 30%)`};
  cursor: pointer;
  background: ${(props) => props.background};

  &:active {
    box-shadow: none;
    padding: 12%;
    transform: ${(props) => `translateY(${props.size})`};
  }
`;

export const OptionIconContainer = styled.div<StyleProps>`
  box-shadow: ${(props) =>
    `inset 0px ${props.size * 0.025}px 0px 0px rgb(0 0 0 / 20%)`};
  background: #fff;
  border-radius: 50%;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;

  &:hover {
    background: #f0f0f0;
  }
`;

export const OptionIcon = styled.img.attrs<StyleProps>((props) => ({
  size: `${props.size * 0.35}px`,
}))<StyleProps>`
  object-fit: contain;
  width: ${(props) => props.size};
  height: ${(props) => props.size};
`;
