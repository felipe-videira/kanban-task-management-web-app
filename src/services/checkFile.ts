export default (path: string): boolean => {
    const http = new XMLHttpRequest();
    http.open('HEAD', path, false);
    http.send();

    return http.status !== 404;
}