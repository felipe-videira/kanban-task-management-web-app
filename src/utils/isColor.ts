export default function isColor(colorStr: string): boolean {
  const s = new Option().style;
  s.color = colorStr;
  return s.color !== "";
}
