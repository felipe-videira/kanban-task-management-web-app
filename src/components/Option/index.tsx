import { InferProps, number } from "prop-types";
import { useMemo } from "react";
import {
  string,
  image,
  color,
  func,
  arrayOf,
  oneOfType,
} from "../../utils/validators";
import { OptionButton, OptionIconContainer, OptionIcon } from "./styles";

interface OnClick {
  (name: string): void;
}

interface GetAltText {
  (name: string): string;
}

function Option(props: InferProps<typeof Option.propTypes>) {
  const alt = useMemo(
    () =>
      props.alt
        ? typeof props.alt === "string"
          ? props.alt
          : props.alt(props.name)
        : "",
    [props.alt, props.name]
  );

  return (
    <OptionButton
      id={props.name}
      type="button"
      onClick={() => props.onClick && props.onClick(props.name)}
      size={props.size}
      background={props.color}
      clickable={!!props.onClick}
    >
      <OptionIconContainer size={props.size}>
        <OptionIcon src={props.icon} alt={alt} size={props.size} />
      </OptionIconContainer>
    </OptionButton>
  );
}

Option.propTypes = {
  name: string.lenght(2).isRequired,
  icon: image.isRequired,
  color: oneOfType([color, arrayOf(color).lenght(2)]).isRequired,
  onClick: func<OnClick>(),
  size: number.isRequired,
  alt: oneOfType([string.lenght(2), func<GetAltText>()]).isRequired,
};

export default Option;
