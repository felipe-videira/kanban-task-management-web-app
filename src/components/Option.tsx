import { InferProps } from "prop-types";
import { useMemo } from "react";
import checkFile from "../services/checkFile";
import customValidator from "../utils/customValidator";
import isColor from "../utils/isColor";
import stringValidator from "../utils/stringValidator";

interface OnClick {
  (name: string): void;
}

function Option({
  name,
  icon,
  color,
  onClick,
}: InferProps<typeof Option.propTypes>) {
  const background = useMemo(
    () =>
      Array.isArray(color) ? `linear-gradient(${color.join(",")})` : color,
    [color]
  );

  return (
    <button type="button" onClick={() => onClick(name)} style={{ background }}>
      <img src={icon} />
    </button>
  );
}

Option.propTypes = {
  name: stringValidator(true, 2),
  icon: customValidator<string>((props, propName): Error | null => {
    if (!props[propName]) {
      return new Error(
        `The prop \`icon\` is marked as required in \`Option\`, but its value is \`${props[propName]}\``
      );
    }
    if (typeof props[propName] !== "string") {
      return new Error(
        `Invalid prop \`icon\` of type \`${typeof props[
          propName
        ]}\` supplied to \`Option\`, expected \`string\``
      );
    }
    if (!checkFile(props[propName])) {
      return new Error(
        `The prop \`icon\` requires a valid path to a image in \`Option\`, but its value is \`${props[propName]}\``
      );
    }

    return null;
  }),
  color: customValidator<string | Array<string>>(
    (props, propName): Error | null => {
      const value = props[propName];
      const isArray = Array.isArray(props[propName]);

      if (!value || (isArray && value.length === 0)) {
        return new Error(
          `The prop \`color\` is marked as required in \`Option\`, but its value is \`${value}\``
        );
      }
      if (
        typeof value !== "string" &&
        (!isArray || !value.every((item: unknown) => typeof item === "string"))
      ) {
        return new Error(
          `Invalid prop \`color\` of type \`${typeof value}\` supplied to \`Option\`, expected \`string\` or \`string[]\``
        );
      }
      if (
        (!isArray && !isColor(value)) ||
        (isArray && !value.every((item: unknown) => isColor(item as string)))
      ) {
        return new Error(
          `The prop \`color\` requires a valid color or array of colors in \`Option\`, but its value is \`${value}\``
        );
      }

      return null;
    }
  ),
  onClick: customValidator<OnClick>((props, propName): Error | null => {
    if (!props[propName]) {
      return new Error(
        `The prop \`onClick\` is marked as required in \`Option\`, but its value is \`${props[propName]}\``
      );
    }

    return null;
  }),
};

export default Option;
