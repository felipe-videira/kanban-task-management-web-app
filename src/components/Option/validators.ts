import checkFile from "../../services/checkFile";
import customValidator from "../../utils/customValidator";
import isColor from "../../utils/isColor";

interface OnClick {
  (name: string): void;
}

export const iconValidator = customValidator<string>((props, propName) => {
  if (!props[propName]) {
    return new Error(
      `The prop \`icon\` is marked as required in \`Option\`, but its value is \`${props[propName]}\``
    );
  }
  if (typeof props[propName] !== "string") {
    return new Error(
      `Invalid prop \`icon\` of type \`${typeof props[
        propName
      ]}\` supplied to \`Option\`, expected \`string\``
    );
  }
  if (!checkFile(props[propName])) {
    return new Error(
      `The prop \`icon\` requires a valid path to a image in \`Option\`, but its value is \`${props[propName]}\``
    );
  }

  return null;
});

export const colorValidator = customValidator<string | Array<string>>(
  (props, propName) => {
    const value = props[propName];
    const isArray = Array.isArray(props[propName]);

    if (!value || (isArray && value.length === 0)) {
      return new Error(
        `The prop \`color\` is marked as required in \`Option\`, but its value is \`${value}\``
      );
    }
    if (
      typeof value !== "string" &&
      (!isArray || !value.every((item: unknown) => typeof item === "string"))
    ) {
      return new Error(
        `Invalid prop \`color\` of type \`${typeof value}\` supplied to \`Option\`, expected \`string\` or \`string[]\``
      );
    }
    if (
      (!isArray && !isColor(value)) ||
      (isArray && !value.every((item: unknown) => isColor(item as string)))
    ) {
      return new Error(
        `The prop \`color\` requires a valid color or array of colors in \`Option\`, but its value is \`${value}\``
      );
    }

    return null;
  }
);

export const onClickValidator = customValidator<OnClick>((props, propName) => {
  if (!props[propName]) {
    return new Error(
      `The prop \`onClick\` is marked as required in \`Option\`, but its value is \`${props[propName]}\``
    );
  }

  return null;
});
