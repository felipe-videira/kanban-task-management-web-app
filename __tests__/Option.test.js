import React from "react";
import { shallow } from "enzyme";
import Option from "../src/components/Option";
import checkFile from "../src/services/checkFile";
import isColor from "../src/utils/isColor";

const mockGameConfig = {
  name: "paper",
  icon: "images/icon-paper.svg",
  colorString: "#000",
  colorArray: ["hsl(230, 89%, 62%)", "hsl(230, 89%, 65%)"],
};
const mockClickFn = jest.fn((name) => {});
const mockInvalidPath = "invalid/path";
const mockInvalidColor = "invalid";

jest.mock("../src/services/checkFile");
jest.mock("../src/utils/isColor");

checkFile.mockImplementation((value) => value !== mockInvalidPath);
isColor.mockImplementation((value) => value !== mockInvalidColor);

describe("Option", () => {
  it("should render correctly with all props, with color as string", () => {
    const component = shallow(
      <Option
        {...mockGameConfig}
        color={mockGameConfig.colorString}
        onClick={mockClickFn}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("should render correctly with all props, with color as array", () => {
    const component = shallow(
      <Option
        {...mockGameConfig}
        color={mockGameConfig.colorArray}
        onClick={mockClickFn}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should throw error if prop "name" is not present', () => {
    expect(() =>
      shallow(
        <Option {...mockGameConfig} name={undefined} onClick={mockClickFn} />
      )
    ).toThrow();
  });

  it('should throw error if prop "name" is not a string', () => {
    expect(() =>
      shallow(<Option {...mockGameConfig} name={1} onClick={mockClickFn} />)
    ).toThrow();
  });

  it('should throw error if prop "name" is below min length', () => {
    expect(() =>
      shallow(<Option {...mockGameConfig} name={"a"} onClick={mockClickFn} />)
    ).toThrow();
  });

  it('should throw error if prop "icon" is not present', () => {
    expect(() =>
      shallow(
        <Option {...mockGameConfig} icon={undefined} onClick={mockClickFn} />
      )
    ).toThrow();
  });

  it('should throw error if prop "icon" is not a string', () => {
    expect(() =>
      shallow(<Option {...mockGameConfig} icon={1} onClick={mockClickFn} />)
    ).toThrow();
  });

  it('should throw error if prop "icon" is not valid', () => {
    expect(() =>
      shallow(
        <Option
          {...mockGameConfig}
          icon={mockInvalidPath}
          onClick={mockClickFn}
        />
      )
    ).toThrow();
  });

  it('should throw error if prop "color" is not present', () => {
    expect(() =>
      shallow(
        <Option {...mockGameConfig} color={undefined} onClick={mockClickFn} />
      )
    ).toThrow();
  });

  it('should throw error if prop "color" is not a string or string[]', () => {
    expect(() =>
      shallow(<Option {...mockGameConfig} color={1} onClick={mockClickFn} />)
    ).toThrow();
  });

  it('should throw error if prop "color" is not valid', () => {
    expect(() =>
      shallow(
        <Option
          {...mockGameConfig}
          color={mockInvalidColor}
          onClick={mockClickFn}
        />
      )
    ).toThrow();
  });

  it('should throw error if prop "onclick" is not present', () => {
    expect(() => shallow(<Option {...mockGameConfig} />)).toThrow();
  });

  it("should call click function", () => {
    const component = shallow(
      <Option {...mockGameConfig} onClick={mockClickFn} />
    );

    component.find("button").simulate("click");

    expect(mockClickFn).toHaveBeenCalled();
  });
});
