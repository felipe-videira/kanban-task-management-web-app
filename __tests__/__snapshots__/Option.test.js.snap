// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option should render correctly with all props, with color as array 1`] = `ShallowWrapper {}`;

exports[`Option should render correctly with all props, with color as string 1`] = `ShallowWrapper {}`;
